File: lib/handler.js
import { fetchImage } from './fetch-image.js';
import { createImageLogger } from './image-logger.js';
import { ParamError, getQueryValue, parseParams } from './parse-params.js';
import { FORMAT_CONTENT_TYPES, probeImageMetadataFromUrl, processImage } from './process-image.js';
import { extractVideoFrameRange, probeVideoMetadataFromUrl } from './process-video.js';
import { buildCacheKey } from './cache.js';

export const CACHE_CONTROL = 'public, max-age=31536000, immutable';
export const PROCESSOR_NAME = 'edge-image';
const MEDIA_ROUTE_PREFIX = '/api/media';
const VIDEO_EXTENSIONS = /\.(mp4|webm)(\?.*)?$/i;

export function createImageHandler({
  fetchImageImpl = fetchImage,
  processImageImpl = processImage,
  probeImageMetadataFromUrlImpl = probeImageMetadataFromUrl,
  probeVideoMetadataFromUrlImpl = probeVideoMetadataFromUrl,
  extractVideoFrameRangeImpl = extractVideoFrameRange,
  logger = console,
  platformConfig,
  cache,
} = {}) {
  return async function imageHandler(req, res) {
    const requestLogger = createImageLogger({
      env: req.env,
      sink: logger,
      requestId: getRequestId(req),
      base: {
        route: '/api/media',
      },
    });

    requestLogger.info('image.request.start', {
      method: req.method || 'GET',
      path: getRequestPath(req),
    });

    if (req.method && req.method !== 'GET') {
      requestLogger.warn('image.request.method_not_allowed', {
        method: req.method,
      });
      return sendJson(res, 405, { error: 'Method Not Allowed' }, { Allow: 'GET' });
    }

    const query = extractQuery(req);
    let params;
    try {
      params = parseParams(extractSourceUrl(req, query), query, { env: req.env, platformConfig });
    } catch (error) {
      if (error instanceof ParamError) {
        requestLogger.warn('image.request.param_error', {
          error,
        });
        return sendJson(res, 400, { error: error.message });
      }

      requestLogger.error('image.request.param_unexpected_error', {
        error,
      });
      return sendJson(res, 500, { error: 'Internal Server Error' });
    }

    const isVideo = VIDEO_EXTENSIONS.test(params.url);

    requestLogger.info('image.request.params', {
      sourceUrl: params.url,
      sourceHost: getUrlHost(params.url),
      width: params.width,
      height: params.height,
      fit: params.fit,
      quality: params.quality,
      format: params.format,
      rotate: params.rotate,
      flip: params.flip || '',
      isVideo,
    });

    if (isVideo && params.format === 'json') {
      // Check metadata cache
      if (cache) {
        const cacheKey = buildCacheKey('meta', params.url);
        const cached = await cache.get(cacheKey);
        if (cached) {
          requestLogger.info('image.request.cache_hit', { type: 'meta' });
          return sendJson(res, 200, JSON.parse(cached.buffer.toString()));
        }
      }

      try {
        const videoMetadata = await probeVideoMetadataFromUrlImpl(params.url, {
          logger: requestLogger,
        });

        const result = {
          width: videoMetadata.width,
          height: videoMetadata.height,
          codec: videoMetadata.codec,
          duration: videoMetadata.duration,
          format: videoMetadata.format,
          sourceUrl: params.url,
          sourceSize: videoMetadata.sourceSize ?? null,
        };

        // Write to metadata cache
        if (cache) {
          const cacheKey = buildCacheKey('meta', params.url);
          await cache.set(cacheKey, { buffer: Buffer.from(JSON.stringify(result)), header: { contentType: 'application/json' } });
        }

        requestLogger.info('image.request.success', { statusCode: 200, videoMetadata });

        return sendJson(res, 200, result);
      } catch (error) {
        requestLogger.warn('image.request.video_probe_failed', {
          error,
          sourceUrl: params.url,
        });

        return sendJson(
          res,
          502,
          {
            error: 'Bad Gateway',
            details: sanitizeHeaderValue(error?.message || 'Video probe failed'),
          },
          { 'X-Processor': PROCESSOR_NAME },
        );
      }
    }

    if (!isVideo && params.format === 'json') {
      // Check metadata cache
      if (cache) {
        const cacheKey = buildCacheKey('meta', params.url);
        const cached = await cache.get(cacheKey);
        if (cached) {
          requestLogger.info('image.request.cache_hit', { type: 'meta' });
          return sendJson(res, 200, JSON.parse(cached.buffer.toString()));
        }
      }

      try {
        const imageMetadata = await probeImageMetadataFromUrlImpl(params.url, {
          logger: requestLogger,
        });

        const result = {
          width: imageMetadata.width,
          height: imageMetadata.height,
          format: imageMetadata.format,
          channels: imageMetadata.channels,
          sourceUrl: params.url,
          sourceContentType: imageMetadata.sourceContentType,
          sourceSize: imageMetadata.sourceSize ?? null,
        };

        // Write to metadata cache
        if (cache) {
          const cacheKey = buildCacheKey('meta', params.url);
          await cache.set(cacheKey, { buffer: Buffer.from(JSON.stringify(result)), header: { contentType: 'application/json' } });
        }

        requestLogger.info('image.request.success', { statusCode: 200, imageMetadata });

        return sendJson(res, 200, result);
      } catch (error) {
        requestLogger.warn('image.request.image_probe_failed', {
          error,
          sourceUrl: params.url,
          sourceHost: getUrlHost(params.url),
        });

        return sendJson(
          res,
          502,
          {
            error: 'Bad Gateway',
            details: sanitizeHeaderValue(error?.message || 'Image metadata probe failed'),
          },
          { 'X-Processor': PROCESSOR_NAME },
        );
      }
    }

    // Check processed result cache
    if (cache) {
      const processedKey = buildCacheKey('processed', params.url, params);
      const cachedProcessed = await cache.get(processedKey);
      if (cachedProcessed) {
        requestLogger.info('image.request.cache_hit', { type: 'processed' });
        const outputContentType = cachedProcessed.header?.contentType || 'application/octet-stream';
        return sendBuffer(res, 200, cachedProcessed.buffer, {
          'Content-Type': outputContentType,
          'Cache-Control': CACHE_CONTROL,
          'X-Processor': PROCESSOR_NAME,
          'X-Cache': 'HIT',
        });
      }
    }

    let imageBuffer;
    let sourceBytes;
    let sourceContentType;

    // Check source cache
    let sourceCacheHit = false;
    if (cache) {
      const sourceKey = buildCacheKey('source', params.url);
      const cachedSource = await cache.get(sourceKey);
      if (cachedSource) {
        imageBuffer = cachedSource.buffer;
        sourceBytes = cachedSource.buffer.length;
        sourceContentType = cachedSource.header?.contentType || 'application/octet-stream';
        sourceCacheHit = true;
        requestLogger.info('image.request.cache_hit', { type: 'source' });
      }
    }

    if (!sourceCacheHit) {
      if (isVideo) {
        try {
          imageBuffer = await extractVideoFrameRangeImpl(params.url, {
            logger: requestLogger,
          });
          sourceBytes = imageBuffer.length;
          sourceContentType = 'image/png';

          // Cache video frame
          if (cache) {
            const sourceKey = buildCacheKey('source', params.url);
            await cache.set(sourceKey, { buffer: imageBuffer, header: { contentType: sourceContentType } });
          }
        } catch (error) {
          requestLogger.warn('image.request.video_frame_failed', {
            error,
            sourceUrl: params.url,
          });

          return sendJson(
            res,
            502,
            {
              error: 'Bad Gateway',
              details: sanitizeHeaderValue(error?.message || 'Video frame extraction failed'),
            },
            { 'X-Processor': PROCESSOR_NAME },
          );
        }
      } else {
        let source;
        try {
          source = await fetchImageImpl(params.url, {
            logger: requestLogger,
          });
        } catch (error) {
          requestLogger.warn('image.request.fetch_failed', {
            error,
            status: error?.status,
            sourceUrl: params.url,
            sourceHost: getUrlHost(params.url),
          });

          return sendJson(
            res,
            502,
            {
              error: 'Bad Gateway',
              details: sanitizeHeaderValue(error?.message || 'Source image fetch failed'),
            },
            { 'X-Processor': PROCESSOR_NAME },
          );
        }

        imageBuffer = source.buffer;
        sourceBytes = source.buffer.length;
        sourceContentType = source.contentType;

        // Cache source image
        if (cache) {
          const sourceKey = buildCacheKey('source', params.url);
          await cache.set(sourceKey, { buffer: imageBuffer, header: { contentType: sourceContentType } });
        }
      }
    }

    try {
      const result = await processImageImpl(
        imageBuffer,
        {
          ...params,
          sourceContentType,
        },
        {
          logger: requestLogger,
        },
      );

      const { buffer, metadata } = result;
      const outputContentType = FORMAT_CONTENT_TYPES[metadata.format] || 'application/octet-stream';

      // Cache processed result
      if (cache) {
        const processedKey = buildCacheKey('processed', params.url, params);
        await cache.set(processedKey, { buffer, header: { contentType: outputContentType, ...metadata } });
      }

      requestLogger.info('image.request.success', {
        statusCode: 200,
        sourceBytes,
        outputBytes: buffer.length,
        outputContentType,
        metadata,
      });

      return sendBuffer(res, 200, buffer, {
        'Content-Type': outputContentType,
        'Cache-Control': CACHE_CONTROL,
        'X-Processor': PROCESSOR_NAME,
        'X-Image-Width': String(metadata.width),
        'X-Image-Height': String(metadata.height),
        'X-Image-Format': metadata.format,
        'X-Image-Size': String(metadata.size),
      });
    } catch (error) {
      requestLogger.warn('image.request.processing_failed_fallback', {
        error,
        sourceBytes,
      });

      return sendBuffer(res, 200, imageBuffer, {
        'Content-Type': sourceContentType || 'application/octet-stream',
        'Cache-Control': CACHE_CONTROL,
        'X-Processor': PROCESSOR_NAME,
        'X-Processing-Error': sanitizeHeaderValue(error?.message || 'Image processing failed'),
      });
    }
  };
}

function extractQuery(req) {
  if (req.query) {
    return req.query;
  }

  const host = req.headers?.host || 'localhost';
  const url = new URL(req.url || '/', `http://${host}`);
  return url.searchParams;
}

function extractSourceUrl(req, query) {
  const encodedPathSource = getEncodedPathSource(req);
  if (encodedPathSource !== undefined) {
    return decodePathSource(encodedPathSource);
  }

  const rewriteSource = getQueryValue(query, 'source');
  if (rewriteSource === undefined || rewriteSource === '') {
    return rewriteSource;
  }

  return normalizeRewriteSource(rewriteSource);
}

function getEncodedPathSource(req) {
  const pathname = getRequestPath(req);
  const sourcePrefix = `${MEDIA_ROUTE_PREFIX}/`;
  if (!pathname.startsWith(sourcePrefix)) {
    return undefined;
  }

  return pathname.slice(sourcePrefix.length);
}

function decodePathSource(encodedPathSource) {
  try {
    return decodeURIComponent(encodedPathSource);
  } catch {
    throw new ParamError('source URL path segment must be valid percent-encoding');
  }
}

function normalizeRewriteSource(source) {
  const value = String(source);
  if (/^https?:\/\//i.test(value)) {
    return value;
  }

  return decodePathSource(value);
}

function sendJson(res, statusCode, body, headers = {}) {
  return sendBuffer(res, statusCode, Buffer.from(JSON.stringify(body)), {
    'Content-Type': 'application/json; charset=utf-8',
    ...headers,
  });
}

function sendBuffer(res, statusCode, body, headers = {}) {
  res.statusCode = statusCode;
  for (const [key, value] of Object.entries(headers)) {
    res.setHeader(key, value);
  }

  res.end(Buffer.isBuffer(body) ? body : Buffer.from(body));
}

export function sanitizeHeaderValue(value) {
  return String(value)
    .replace(/[^\x20-\x7E]/g, ' ')
    .slice(0, 180);
}

function getRequestId(req) {
  return (
    getHeader(req, 'x-vercel-id') ||
    getHeader(req, 'x-request-id') ||
    getHeader(req, 'x-correlation-id') ||
    undefined
  );
}

function getHeader(req, name) {
  const headers = req.headers || {};
  if (typeof headers.get === 'function') {
    return headers.get(name) || undefined;
  }

  const normalized = name.toLowerCase();
  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() === normalized) {
      return Array.isArray(value) ? value[0] : value;
    }
  }

  return undefined;
}

function getRequestPath(req) {
  try {
    return new URL(req.url || '/api/media', 'http://localhost').pathname;
  } catch {
    return req.url || '/api/media';
  }
}

function getUrlHost(url) {
  try {
    return new URL(url).host;
  } catch {
    return '';
  }
}

File: internal/server/handler.go
package server

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"regexp"
	"strings"

	"edge-image/internal/allowlist"
	"edge-image/internal/cache"
	"edge-image/internal/config"
	"edge-image/internal/fetcher"
	"edge-image/internal/params"
	"edge-image/internal/processor"
)

const (
	cacheControl  = "public, max-age=31536000, immutable"
	processorName = "edge-image"
	mediaRoute    = "/api/media/"
)

var videoExtensions = regexp.MustCompile(`\.(mp4|webm)(\?.*)?$`)

type Handler struct {
	cfg   config.PlatformConfig
	al    *allowlist.Allowlist
	cache *cache.Cache
	fetch *fetcher.Fetcher
	log   *slog.Logger
}

func NewHandler(cfg config.PlatformConfig, al *allowlist.Allowlist, c *cache.Cache, f *fetcher.Fetcher, log *slog.Logger) *Handler {
	return &Handler{cfg: cfg, al: al, cache: c, fetch: f, log: log}
}

func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.Header().Set("Allow", "GET")
		sendJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "Method Not Allowed"})
		return
	}

	sourceURL := extractSourceURL(r)
	if sourceURL == "" {
		sendJSON(w, http.StatusBadRequest, map[string]string{"error": "Missing source URL"})
		return
	}

	p, err := params.Parse(sourceURL, r.URL.Query(), h.cfg, h.al)
	if err != nil {
		sendJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
		return
	}

	isVideo := videoExtensions.MatchString(p.URL)
	log := h.log.With("sourceUrl", p.URL, "isVideo", isVideo)

	if p.Format == "json" {
		h.handleMetadata(w, p, isVideo, log)
		return
	}

	// Check processed cache
	if h.cache != nil {
		key := buildProcessedCacheKey(p)
		if entry, found := h.cache.Get(key); found {
			w.Header().Set("Content-Type", entry.ContentType)
			w.Header().Set("Cache-Control", cacheControl)
			w.Header().Set("X-Processor", processorName)
			w.Header().Set("X-Cache", "HIT")
			w.Write(entry.Buffer)
			return
		}
	}

	var sourceBuffer []byte
	var sourceContentType string

	if isVideo {
		frame, err := processor.ExtractVideoFrame(p.URL, h.fetch, log)
		if err != nil {
			log.Warn("handler: video frame extraction failed", "error", err)
			sendJSON(w, http.StatusBadGateway, map[string]string{
				"error":   "Bad Gateway",
				"details": sanitizeHeader(err.Error()),
			})
			return
		}
		sourceBuffer = frame
		sourceContentType = "image/png"
	} else {
		sourceCached := false
		if h.cache != nil {
			sourceKey := cache.BuildCacheKey("source", p.URL)
			if entry, found := h.cache.Get(sourceKey); found {
				sourceBuffer = entry.Buffer
				sourceContentType = entry.ContentType
				sourceCached = true
			}
		}

		if !sourceCached {
			result, err := h.fetch.FetchImage(p.URL)
			if err != nil {
				log.Warn("handler: fetch failed", "error", err)
				sendJSON(w, http.StatusBadGateway, map[string]string{
					"error":   "Bad Gateway",
					"details": sanitizeHeader(err.Error()),
				})
				return
			}
			sourceBuffer = result.Buffer
			sourceContentType = result.ContentType

			if h.cache != nil {
				sourceKey := cache.BuildCacheKey("source", p.URL)
				h.cache.Set(sourceKey, &cache.Entry{
					Buffer:      sourceBuffer,
					ContentType: sourceContentType,
				})
			}
		}
	}

	procParams := processor.ImageParams{
		Width:             p.Width,
		Height:            p.Height,
		Fit:               p.Fit,
		Quality:           p.Quality,
		Format:            p.Format,
		Rotate:            p.Rotate,
		Flip:              p.Flip,
		Background:        p.Background,
		MaxDimension:      p.MaxDimension,
		SourceContentType: sourceContentType,
	}

	result, err := processor.ProcessImage(sourceBuffer, procParams, log)
	if err != nil {
		log.Warn("handler: processing failed, returning original", "error", err)
		w.Header().Set("Content-Type", sourceContentType)
		w.Header().Set("Cache-Control", cacheControl)
		w.Header().Set("X-Processor", processorName)
		w.Header().Set("X-Processing-Error", sanitizeHeader(err.Error()))
		w.Write(sourceBuffer)
		return
	}

	outputContentType := processor.FormatContentTypes[result.Format]
	if outputContentType == "" {
		outputContentType = "application/octet-stream"
	}

	if h.cache != nil {
		key := buildProcessedCacheKey(p)
		h.cache.Set(key, &cache.Entry{
			Buffer:      result.Buffer,
			ContentType: outputContentType,
		})
	}

	w.Header().Set("Content-Type", outputContentType)
	w.Header().Set("Cache-Control", cacheControl)
	w.Header().Set("X-Processor", processorName)
	w.Header().Set("X-Image-Width", fmt.Sprintf("%d", result.Width))
	w.Header().Set("X-Image-Height", fmt.Sprintf("%d", result.Height))
	w.Header().Set("X-Image-Format", result.Format)
	w.Header().Set("X-Image-Size", fmt.Sprintf("%d", result.Size))
	w.Write(result.Buffer)
}

func (h *Handler) handleMetadata(w http.ResponseWriter, p *params.Params, isVideo bool, log *slog.Logger) {
	if h.cache != nil {
		key := cache.BuildCacheKey("meta", p.URL)
		if entry, found := h.cache.Get(key); found {
			w.Header().Set("Content-Type", "application/json; charset=utf-8")
			w.Header().Set("Cache-Control", cacheControl)
			w.Header().Set("X-Processor", processorName)
			w.Write(entry.Buffer)
			return
		}
	}

	var data interface{}

	if isVideo {
		meta, err := processor.ProbeVideoMetadata(p.URL, h.fetch, log)
		if err != nil {
			log.Warn("handler: video probe failed", "error", err)
			sendJSON(w, http.StatusBadGateway, map[string]string{
				"error":   "Bad Gateway",
				"details": sanitizeHeader(err.Error()),
			})
			return
		}
		data = map[string]interface{}{
			"width":      meta.Width,
			"height":     meta.Height,
			"codec":      meta.Codec,
			"duration":   meta.Duration,
			"format":     meta.Format,
			"sourceUrl":  p.URL,
			"sourceSize": meta.SourceSize,
		}
	} else {
		// Check source cache first — a prior image processing request may have cached the full source bytes
		var sourceBuffer []byte
		var sourceContentType string
		if h.cache != nil {
			sourceKey := cache.BuildCacheKey("source", p.URL)
			if entry, found := h.cache.Get(sourceKey); found {
				sourceBuffer = entry.Buffer
				sourceContentType = entry.ContentType
			}
		}

		if sourceBuffer == nil {
			result, err := h.fetch.FetchImageRange(p.URL, fetcher.DefaultMetaRange)
			if err != nil {
				log.Warn("handler: image metadata fetch failed", "error", err)
				sendJSON(w, http.StatusBadGateway, map[string]string{
					"error":   "Bad Gateway",
					"details": sanitizeHeader(err.Error()),
				})
				return
			}
			sourceBuffer = result.Buffer
			sourceContentType = result.ContentType
		}

		meta, err := processor.ProbeImageMetadata(sourceBuffer, log)
		if err != nil {
			if len(sourceBuffer) >= int(fetcher.DefaultMetaRange) {
				// Partial download probe failed — try full download
				log.Warn("handler: image probe failed, trying full download", "error", err)
				fullResult, fullErr := h.fetch.FetchImage(p.URL)
				if fullErr != nil {
					sendJSON(w, http.StatusBadGateway, map[string]string{
						"error":   "Bad Gateway",
						"details": sanitizeHeader(fullErr.Error()),
					})
					return
				}
				sourceBuffer = fullResult.Buffer
				sourceContentType = fullResult.ContentType
				meta, err = processor.ProbeImageMetadata(sourceBuffer, log)
				if err != nil {
					sendJSON(w, http.StatusBadGateway, map[string]string{
						"error":   "Bad Gateway",
						"details": sanitizeHeader(err.Error()),
					})
					return
				}
			} else {
				sendJSON(w, http.StatusBadGateway, map[string]string{
					"error":   "Bad Gateway",
					"details": sanitizeHeader(err.Error()),
				})
				return
			}
		}
		meta.SourceSize = int64(len(sourceBuffer))
		meta.SourceContentType = sourceContentType

		data = map[string]interface{}{
			"width":             meta.Width,
			"height":            meta.Height,
			"format":            meta.Format,
			"channels":          meta.Channels,
			"sourceUrl":         p.URL,
			"sourceContentType": meta.SourceContentType,
			"sourceSize":        meta.SourceSize,
		}
	}

	jsonData, _ := json.Marshal(data)

	if h.cache != nil {
		key := cache.BuildCacheKey("meta", p.URL)
		h.cache.Set(key, &cache.Entry{
			Buffer:      jsonData,
			ContentType: "application/json",
		})
	}

	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Header().Set("Cache-Control", cacheControl)
	w.Header().Set("X-Processor", processorName)
	w.Write(jsonData)
}

func extractSourceURL(r *http.Request) string {
	path := r.URL.Path
	if strings.HasPrefix(path, mediaRoute) {
		encoded := path[len(mediaRoute):]
		if encoded == "" {
			return ""
		}
		decoded, err := decodePercentEncoding(encoded)
		if err != nil {
			return ""
		}
		return decoded
	}

	source := r.URL.Query().Get("source")
	if source != "" {
		if strings.HasPrefix(source, "http://") || strings.HasPrefix(source, "https://") {
			return source
		}
		decoded, err := decodePercentEncoding(source)
		if err != nil {
			return ""
		}
		return decoded
	}

	return ""
}

func decodePercentEncoding(s string) (string, error) {
	var result strings.Builder
	for i := 0; i < len(s); i++ {
		if s[i] == '%' {
			if i+2 >= len(s) {
				return "", fmt.Errorf("invalid percent encoding")
			}
			high := hexDigit(s[i+1])
			low := hexDigit(s[i+2])
			if high < 0 || low < 0 {
				return "", fmt.Errorf("invalid percent encoding")
			}
			result.WriteByte(byte(high*16 + low))
			i += 2
		} else {
			result.WriteByte(s[i])
		}
	}
	return result.String(), nil
}

func hexDigit(c byte) int {
	switch {
	case c >= '0' && c <= '9':
		return int(c - '0')
	case c >= 'A' && c <= 'F':
		return int(c-'A') + 10
	case c >= 'a' && c <= 'f':
		return int(c-'a') + 10
	default:
		return -1
	}
}

func sanitizeHeader(value string) string {
	var result strings.Builder
	for _, r := range value {
		if r >= 32 && r <= 126 {
			result.WriteRune(r)
		} else {
			result.WriteRune(' ')
		}
		if result.Len() >= 180 {
			break
		}
	}
	return result.String()
}

func sendJSON(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(data)
}

func buildProcessedCacheKey(p *params.Params) string {
	return cache.BuildCacheKey("processed", p.URL,
		fmt.Sprintf("%d", p.Width), fmt.Sprintf("%d", p.Height),
		p.Fit, fmt.Sprintf("%d", p.Quality), p.Format,
		fmt.Sprintf("%d", p.Rotate), p.Flip)
}
